### PR TITLE
Extract the truncate string logic to a new helper class.

### PR DESCRIPTION
**Pull request by mesmacosta**
**- What I did**
Created the `datacatalog_strings_helper.py` class to centralize helper string methods.

**- How I did it**
Created the `datacatalog_strings_helper.py` class to centralize helper string methods.

**- How to verify it**
At the moment this is used only on string tag fields, so create a string tag field with more than 2000 bytes, when running the connectors.

**- Description for the changelog**
This truncate string logic is tied to the tag field logic, but this is needed in other places of the connectors. Testing the hive connector with large table names, we found out that the linkedResource can also only have 200 bytes.

Example:
```
google.api_core.exceptions.InvalidArgument: 400 "//35.223.225.48//hdfs://namenode:8020/user/hive/wa..." is an invalid value for CreateEntryRequest.entry.linked_resource name. It must: contain only letters, numbers, periods, colons, slashs, underscores, dashes and hashes and be at most 200  bytes long when encoded in UTF-8.
```

To sum up, this logic is needed in other places and makes sense to decouple it from the tag field factory. Also the `max_length` needs to be parametrized  since the linkedResource limit is different from the string tag field.

### CODE REVIEW

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 from google.protobuf import timestamp_pb2
 import six
 
+from google.datacatalog_connectors.commons import prepare
+
 
 class BaseTagFactory:
     __UTF8_CHARACTER_ENCODING = 'UTF-8'
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py:25]: Are all these fields still needed here?

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# limitations under the License.
+
+import six
+
+
+class DataCatalogStringsHelper:
+    __UTF8_CHARACTER_ENCODING = 'UTF-8'
+    # String field values are limited to 2000 bytes size when encoded in UTF-8.
+    __STRING_VALUE_UTF8_MAX_LENGTH = 2000
+    __SUFFIX_CHARS_LENGTH = 3
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:24]:
```suggestion
    __TRUNCATED_STRING_SUFFIX_CHARS_LENGTH = 3
```

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import six
+
+
+class DataCatalogStringsHelper:
+    __UTF8_CHARACTER_ENCODING = 'UTF-8'
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:21]: Please sort the constants alphabetically.

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        """
+        if not (value is not None and isinstance(value, six.string_types)):
+            return
+
+        encoding = cls.__UTF8_CHARACTER_ENCODING
+        suffix_length = cls.__SUFFIX_CHARS_LENGTH
+
+        encoded = value.encode(encoding)
+
+        # the max length supported is stored at max_length
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:58]:
```suggestion
        # The max length supported is stored at max_length.
```

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        # we leave some chars as the suffix_length to be used when
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:59]:
```suggestion
        # We leave some chars as the suffix_length to be used when
```

File: google-datacatalog-connectors-commons/tests/google/datacatalog_connectors/commons/prepare/base_tag_factory_test.py [hunk 6]
```diff
@@ -18,13 +18,16 @@
 import calendar
 import unittest
 
-from google.datacatalog_connectors.commons import prepare
 import dateutil.parser
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/tests/google/datacatalog_connectors/commons/prepare/base_tag_factory_test.py:21]:
```suggestion
from dateutil import parser
```
Makes sense?

File: google-datacatalog-connectors-commons/tests/google/datacatalog_connectors/commons/prepare/base_tag_factory_test.py [hunk 7]
```diff
@@ -18,13 +18,16 @@
...
-
+import mock
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/tests/google/datacatalog_connectors/commons/prepare/base_tag_factory_test.py:22]: Please sort this import block alphabetically.

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+          Cyrillic, Coptic, Armenian, Hebrew, Arabic, Syriac, Thaana and N'Ko
+          alphabets, as well as Combining Diacritical Marks;
+        - three bytes are needed for characters in the rest of the Basic
+          Multilingual Plane, which contains virtually all characters in common
+          use, including most Chinese, Japanese and Korean characters;
+        - four bytes are needed for characters in the other planes of Unicode,
+          which include less common CJK characters, various historic scripts,
+          mathematical symbols, and emoji (pictographic symbols).
+
+        Given a value and a string Tag Field, this method assigns the field the
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:44]: Please review this docstring. Mentioning a Tag Field here is out of context. Some parts should be kept in the former `_set_string_field` method, others should actually be moved to the refactored code.

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import six
+
+
+class DataCatalogStringsHelper:
+    __UTF8_CHARACTER_ENCODING = 'UTF-8'
+    # String field values are limited to 2000 bytes size when encoded in UTF-8.
+    __STRING_VALUE_UTF8_MAX_LENGTH = 2000
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:23]: This is the max length for Tag Fields. Since the refactored class is intended to serve multiple purposes, I suggest always asking for the max length to be considered, which means no constant value here and no default value in the `truncate_string` method. WDYT? 🤔

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        suffix_length = cls.__SUFFIX_CHARS_LENGTH
+
+        encoded = value.encode(encoding)
+
+        # the max length supported is stored at max_length
+        # we leave some chars as the suffix_length to be used when
+        # creating the new string, so this line truncates the existing string.
+        truncated_string_field = encoded[:max_length - suffix_length]
+
+        decoded = u'{}...'.format(
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:63]:
```suggestion
        decoded = u'{}...'.format(truncated_string_field.decode(encoding, 'ignore')) \
            if len(encoded) > max_length \
            else encoded.decode(encoding, 'ignore')
```
WDYT about enforcing such a manually formatted code here to improve readability?

**Review by ricardolsmendes — CHANGES_REQUESTED** [review]
Hey @mesmacosta, please take a look at the comments I've just left. 🏓

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py [hunk 11]
```diff
@@ -18,12 +18,11 @@
 from google.protobuf import timestamp_pb2
 import six
 
+from google.datacatalog_connectors.commons import prepare
+
 
 class BaseTagFactory:
-    __UTF8_CHARACTER_ENCODING = 'UTF-8'
-    # String field values are limited to 2000 bytes size when encoded in UTF-8.
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py:24]: This comment refers to the `__STRING_VALUE_UTF8_MAX_LENGTH`, so it shouldn't be deleted.

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import six
+
+
+class DataCatalogStringsHelper:
+    # String field values are limited to 2000 bytes size when encoded in UTF-8.
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:21]: Please delete this comment.

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+          Cyrillic, Coptic, Armenian, Hebrew, Arabic, Syriac, Thaana and N'Ko
+          alphabets, as well as Combining Diacritical Marks;
+        - three bytes are needed for characters in the rest of the Basic
+          Multilingual Plane, which contains virtually all characters in common
+          use, including most Chinese, Japanese and Korean characters;
+        - four bytes are needed for characters in the other planes of Unicode,
+          which include less common CJK characters, various historic scripts,
+          mathematical symbols, and emoji (pictographic symbols).
+
+        Given a string, this method checks the value's UTF-8 byte-size and
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:43]:
```suggestion
        Given a string, this method checks its UTF-8 byte-size and
```

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+import six
+
+
+class DataCatalogStringsHelper:
+    # String field values are limited to 2000 bytes size when encoded in UTF-8.
+    __TRUNCATED_STRING_SUFFIX_CHARS_LENGTH = 3
+    __UTF8_CHARACTER_ENCODING = 'UTF-8'
+
+    @classmethod
+    def truncate_string(cls, value, max_length):
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:26]:
```suggestion
    def truncate_string(cls, string, max_length):
```

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py [hunk 15]
```diff
@@ -41,50 +40,15 @@ def _set_double_field(cls, tag, field_id, value):
 
     @classmethod
     def _set_string_field(cls, tag, field_id, value):
-        """
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py:44]: WDYT about leaving the following docstring here?
```
"""
String field values are limited by Data Catalog API at 2000 chars
length when encoded in UTF-8. Given a string Tag Field and its
value, this method assigns the value to the field, truncating if needed.
"""
```

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        """
+        String field values are limited by Data Catalog API at 2000 chars
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:28]: Please delete the first phrase as it only matters to the former `_set_string_field` method.

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py (hunk 13 above)
Comment (ricardolsmendes) [reply google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper.py:43]: Maybe it makes sense, for the sake of clarity, to exchange the order of these paragraphs. WDYT? 🤔

File: google-datacatalog-connectors-commons/tests/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper_test.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertEqual(2000, len(truncated_string.encode('UTF-8')))
+
+    def test_truncate_string_should_truncate_uft8_bytes_size_i18n(self):
+        """
+         - Input string: 1010 'ã' chars;
+         - Expected field value: '998 "ã" chars + ...' since each 'ã' char
+         needs 2 bytes and periods need 1 byte when encoded in UTF-8.
+        """
+
+        str_value = u''
```
Comment (ricardolsmendes) [inline google-datacatalog-connectors-commons/tests/google/datacatalog_connectors/commons/prepare/datacatalog_strings_helper_test.py:64]: The `_value` suffix has no meaning in the refactored code, here and in the next methods. WDYT about renaming this and next vars to simply `string` or something similar?

**Review by ricardolsmendes — CHANGES_REQUESTED** [review]
Thanks, @mesmacosta. I've left a few more comments. 🏓

File: google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py (hunk 11 above)
Comment (mesmacosta) [reply google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py:24]: The constant name and the comment are basically the same. Do you really think that comment helps?

Comment (ricardolsmendes) [reply google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_tag_factory.py:24]: Makes sense, it can be deleted. 👍🏼

**Review by ricardolsmendes — APPROVED** [review]
Great work, @mesmacosta! Thanks for working on this refactoring.
